File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
// @Author: KING-SEN

class Solution {
private:
    vector<vector<int>> dp;
public:
    int lcs(int p1, int p2, string& text1, string& text2){
        if(p1 < 0 || p2 < 0){
            return 0;
        }
        int &ret = dp[p1][p2];
        if(ret != -1){
            return ret;
        }
        if(text1[p1] == text2[p2]){
            ret = 1 + lcs(p1 - 1, p2 - 1, text1, text2);
        }
        else{
            ret = max(lcs(p1 - 1, p2, text1, text2), lcs(p1, p2 - 1, text1, text2));
        }
        return ret;
    }
    int longestCommonSubsequence(string text1, string text2) {
        int len1 = text1.size(), len2 = text2.size();
        dp.resize(len1 + 1, vector<int>(len2, -1));
        return lcs(len1 - 1, len2 - 1, text1, text2);
    }
};

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def func(self, i: int, j: int, s1: str, s2: str, dp: list()) -> int:
        if i == len(s1):
            return 0
        if j == len(s2):
            return 0
        if dp[i][j] != -1:
            return dp[i][j]
        res = float('-inf')
        if s1[i] == s2[j]:
            res = 1 + self.func(i + 1, j + 1, s1, s2, dp)
        else:
            res = max(res, self.func(i + 1, j, s1, s2, dp))
            res = max(res, self.func(i, j + 1, s1, s2, dp))
        dp[i][j] = res
        return res
    
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp = [[-1] * (len(text2) + 1) for _ in range(len(text1) + 1)]
        return self.func(0, 0, text1, text2, dp)